File: [69]_x-的平方根/69.x-的平方根.java
/*
 * @lc app=leetcode.cn id=69 lang=java
 * @lcpr version=21703
 *
 * [69] x 的平方根 
 */

// @lc code=start
class Solution {
    public int mySqrt(int x) {
        if (x == 0) {
            return 0;
        }

        double C = x, x0 = x;
        while (true) {
            double xi = 0.5 * (x0 + C / x0);
            if (Math.abs(x0 - xi) < 1e-7) {
                break;
            }
            x0 = xi;
        }
        return (int) x0;
    }
}
// @lc code=end

// @lcpr-div-debug-arg-start
// funName=
// paramTypes= []
// returnType=
// @lcpr-div-debug-arg-end

/*
 * // @lcpr case=start
 * // 4\n
 * // @lcpr case=end
 * 
 * // @lcpr case=start
 * // 8\n
 * // @lcpr case=end
 * 
 */

File: [69]_x-的平方根/69.x-的平方根.c
/*
 * @lc app=leetcode.cn id=69 lang=c
 *
 * [69] x 的平方根 
 */

// @lc code=start


int mySqrt(int x){
    long i; 
    float x2, y; 
    const float threehalfs = 1.5F ;
    x2 = x * 0.5F ; 
    y = x ; 
    i = * ( long * ) &y; // evil floating point bit level hacking 
    i = 0x5f3759df - ( i >> 1 ); // what the fuck? 
    y = * ( float * ) &i; 
    y = y * ( threehalfs - ( x2 * y * y ) ); // 1st iteration 
//  y = y * ( threehalfs - ( x2 * y * y ) ); // 2nd iteration, this can be removed
    return y ;
}
// @lc code=end

